src/audio/volume/volume.c

```diff
@@ -390,31 +380,161 @@ static struct comp_dev *volume_new(const struct comp_driver *drv,
+	if (curve_type == IPC4_AUDIO_CURVE_TYPE_WINDOWS_FADE) {
+		cd->ramp_type = SOF_VOLUME_WINDOWS_FADE;
+		cd->ramp_func =  &volume_windows_fade_ramp;
+	}
+#endif
+	return 0;
+}
+
+static int init_volume_ipc4(struct comp_dev *dev,
+			    struct comp_ipc_config *config, void *spec)
+{
+	struct ipc4_peak_volume_module_cfg *vol = spec;
+	struct vol_data *cd = comp_get_drvdata(dev);
+	uint32_t channels_count;
+	uint8_t channel_cfg;
+	uint8_t channel;
+
+	mailbox_hostbox_read(&cd->base, sizeof(cd->base), 0, sizeof(cd->base));
+
+	channels_count = cd->base.audio_fmt.channels_count;
+
+	for (channel = 0; channel < channels_count ; channel++) {
+		if (vol->config[0].channel_id == IPC4_ALL_CHANNELS_MASK)
+			channel_cfg = 0;
+		else
+			channel_cfg = channel;
+
+		/* In IPC4 driver sends volume in Q1.31 format. It is converted
+		 * into Q1.23 format to be processed by firmware.
+		 */
+		vol->config[channel].target_volume = Q_SHIFT_RND(vol->config[channel].target_volume,
+								 31, 23);
+
+		set_volume_ipc4(cd, channel,
+				vol->config[channel_cfg].target_volume,
+				vol->config[channel_cfg].curve_type,
+				vol->config[channel_cfg].curve_duration);
+	}
+
+	/* In IPC4 driver sends curve_duration in hundred of ns - it should be
+	 * converted into ms value required by firmware
+	 */
+	cd->initial_ramp  = Q_MULTSR_32X32(vol->config[0].curve_duration,
+					   Q_CONVERT_FLOAT(1.0 / 10000, 31), 0, 31, 0);
+
+	if (!cd->initial_ramp) {
+		/* In case when initial ramp time is equal to zero, vol_min and
+		 * vol_max variables should be set to target_volume value
+		 */
+		cd->vol_min = vol->config[0].target_volume;
+		cd->vol_max = vol->config[0].target_volume;
+	} else {
+		cd->vol_min = VOL_MIN;
+		cd->vol_max = VOL_MAX;
 	}
 
+	cd->mailbox_offset = offsetof(struct ipc4_fw_registers, peak_vol_regs);
+	cd->mailbox_offset += IPC4_INST_ID(config->id) * sizeof(struct ipc4_peak_volume_regs);
+
 	reset_state(cd);
-	comp_info(dev, "vol->initial_ramp = %d, vol->ramp = %d, vol->min_value = %d, vol->max_value = %d",
-		  vol->initial_ramp, vol->ramp,
-		  vol->min_value, vol->max_value);
+
+	return 0;
+}
+#endif /* CONFIG_IPC_MAJOR_4 */
+
+/**
+ * \brief Creates volume component.
+ *
+ * \return Pointer to volume base component device.
+ */
+static struct comp_dev *volume_new(const struct comp_driver *drv,
+				   struct comp_ipc_config *config,
+				   void *spec)
+{
+	struct comp_dev *dev;
+	struct vol_data *cd;
+	int ret;
+
+	comp_cl_dbg(&comp_volume, "volume_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	dev->ipc_config = *config;
+	list_init(&dev->bsource_list);
+	list_init(&dev->bsink_list);
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+		     sizeof(struct vol_data));
+	if (!cd)
+		goto fail;
+
+	comp_set_drvdata(dev, cd);
+
+#if CONFIG_IPC_MAJOR_3
+	ret = init_volume_ipc3(dev, config, spec);
+#elif CONFIG_IPC_MAJOR_4
+	ret = init_volume_ipc4(dev, config, spec);
+#endif
```

Comment:
To simplify we just need to do more of
```c
#if IPC3
int init_volume(dev, config, spec)
{
....
}
#elif IPC4
int init_volume(dev, config, spec)
{
....
}
#endif
```
This way all our clients dont need the conditional logic.

Reply:
@lgirdwood I am wondering if it makes sense to have ops within the comp specific vol_data so that during init, we can do all the #if IPC3 elseif IPC4 once and set everything instead of having so many #if's everywhere all over the code. What do you think?

Reply:
Yes, this is what I was meaning - we minimize the conditional logic to one large block using common API calls. This way all the client code is simplified and the differentiation is in the conditional block only (not everywhere in the file).

Reply:
Updated with init_volume change